src/java/org/apache/cassandra/db/marshal/ReversedType.java

```diff
@@ -35,25 +39,48 @@
 
 public class ReversedType<T> extends AbstractType<T>
 {
+    private final static Logger logger = LoggerFactory.getLogger(ReversedType.class);
+
     // interning instances
     private static final Map<AbstractType<?>, ReversedType> instances = new ConcurrentHashMap<>();
 
     public final AbstractType<T> baseType;
 
-    public static <T> ReversedType<T> getInstance(TypeParser parser)
+    public static AbstractType<?> getInstance(TypeParser parser)
     {
         List<AbstractType<?>> types = parser.getTypeParameters();
         if (types.size() != 1)
             throw new ConfigurationException("ReversedType takes exactly one argument, " + types.size() + " given");
-        return getInstance((AbstractType<T>) types.get(0));
+        return getInstance(types.get(0));
     }
 
-    public static <T> ReversedType<T> getInstance(AbstractType<T> baseType)
+    public static <T> AbstractType<T> getInstance(AbstractType<T> baseType)
     {
-        ReversedType<T> t = instances.get(baseType);
-        return null == t
-             ? instances.computeIfAbsent(baseType, ReversedType::new)
-             : t;
+        ReversedType<T> type = instances.get(baseType);
+        if (type != null)
+            return type;
+
+        // Stacking ReversedType is really useless and would actually break some of the code (typically,
+        // AbstractType#isValueCompatibleWith would end up hitting the exception thrown by
+        // ReversedType#isValueCompatibleWithInternal). So we should throw if we find such stacking. But at the
+        // same time, we can't be 100% no-one ever made that mistake somewhere, and it went un-noticed, and it's
+        // probably not worth risking breaking them. So we log an error but otherwise "cancel-out" the double
+        // reverse.
+        if (baseType instanceof ReversedType<?>)
+        {
+            // Note: users have no way to input ReversedType outside custom types (which are pretty
+            // confidential in the first place), so if this is ever triggered, this will likely be an internal.
+            // So shipping an exception in the logger output to get a stack trace and make debugging easier.
+            logger.error("Detected a type with 2 ReversedType() back-to-back, which is not allowed. This should "
+                         + "be looked at as this is likely unintended, but cancelling out the double reversion in "
+                         + "the meantime", new RuntimeException("Invalid double-reversion"));
+            return ((ReversedType<T>) baseType).baseType;
```

Comment:
We have removed the automatic type fixing for multicell subtypes in `AbstractType`, but we still have automatic fixing here. I think we should have the same behaviour in both, whether it's fixing or not.